### PR TITLE
Refactor hopp solver

### PR DESCRIPTION
**Pull request by bayc**
# Refactor HOPP solver to be more modular

This PR generalizes the HOPP dispatch structure to more easily allow for the inclusion of new technologies and facilitate debugging of the dispatch codes. This will also support adding new dispatch strategies.

## Related issue
#307 
#308 

## Impacted areas of the software
The dispatch classes and code.

## Additional supporting information
Most of the changes are moving code from `hybrid_dispatch.py` into the respective technology dispatch classes.

Other significant changes are calling general functions from `HybridDispatch` such as `self.power_sources[tech]._dispatch._create_variables()` while looping over technologies instead of building the attribute to be retrieved with through strings. This will make debugging easier.

The remaining changes are mainly formatting the code for improved readability.

## Test results, if applicable
The tests will be passing.

### CODE REVIEW

File: hopp/simulation/technologies/dispatch/hybrid_dispatch.py [hunk 1]
```diff
@@ -71,147 +63,51 @@ def _create_parameters(hybrid):
...
-
-    def _create_grid_port(self, hybrid, t):
-        hybrid.grid_port = Port(initialize={'system_generation': hybrid.system_generation,
-                                            'system_load': hybrid.system_load,
-                                            'electricity_sold': hybrid.electricity_sold,
-                                            'electricity_purchased': hybrid.electricity_purchased})
-        self.ports[t].append(hybrid.grid_port)
-
-    def _create_grid_constraints(self, hybrid, t):
+    def _create_variables_and_ports(self, hybrid, t):
```
Comment (kbrunik) [inline hopp/simulation/technologies/dispatch/hybrid_dispatch.py:69]: I'd love to see a doc string for this to explain briefly how this goes about setting up the variables and ports for Pyomo

File: hopp/simulation/technologies/dispatch/hybrid_dispatch.py [hunk 2]
```diff
@@ -244,71 +140,25 @@ def _delete_objective(self):
     def create_max_gross_profit_objective(self):
```
Comment (kbrunik) [inline hopp/simulation/technologies/dispatch/hybrid_dispatch.py:140]: Add docstring

File: hopp/simulation/technologies/dispatch/hybrid_dispatch.py [hunk 3]
```diff
@@ -319,61 +169,22 @@ def gross_profit_objective_rule(m):
     def create_min_operating_cost_objective(self):
```
Comment (kbrunik) [inline hopp/simulation/technologies/dispatch/hybrid_dispatch.py:169]: Add docstring

**Review by kbrunik — COMMENTED** [review]
I think it looks really great in general. I do think that while we go about refactoring different pieces of the solver that it would be valuable to add docstrings. 

At a minimum I'd like to see a docstring for each individual technology dispatch and for each objective function.

I think it would be valuable to add docstrings for the functions you've added/changed even internal functions to help make it easier for future developers to add additional technologies and start to understand the various pyomo connections.

File: hopp/simulation/technologies/dispatch/power_sources/wave_dispatch.py [hunk 4]
```diff
@@ -1,23 +1,67 @@
...
+            blocks[t].time_weighting_factor 
+            * self.blocks[t].time_duration
+            * self.blocks[t].cost_per_generation
+            * blocks[t].wave_generation
+            for t in blocks.index_set()
+        )
+
+    def _create_variables(self, hybrid):
+        hybrid.wave_generation = Var(
+            doc="Power generation of wind turbines [MW]",
```
Comment (jaredthomas68) [inline hopp/simulation/technologies/dispatch/power_sources/wave_dispatch.py:58]: Doc is incorrect. We need to fix this typo

**Review by jaredthomas68 — APPROVED** [review]
I agree with Kaitlin. This looks fine, but some doc strings would go a long way. Since you are up against a milestone I'm fine with merging it in now, but at least make an issue to add the doc strings. I also pointed out one typo that should be fixed prior to merge.

File: hopp/simulation/technologies/dispatch/grid_dispatch.py [hunk 5]
```diff
@@ -35,6 +42,86 @@ def dispatch_block_rule(self, grid):
         # Ports
         self._create_grid_ports(grid)
 
+    def max_gross_profit_objective(self, blocks):
+        self.obj = pyomo.Expression(
```
[the remark this answers is not in the document]
Comment (bayc) [reply hopp/simulation/technologies/dispatch/grid_dispatch.py:46]: I did, but I was unsure of knock-on effects. As we discussed, we will consider this in the future.

File: hopp/simulation/technologies/dispatch/power_sources/wave_dispatch.py (hunk 4 above)
Comment (bayc) [reply hopp/simulation/technologies/dispatch/power_sources/wave_dispatch.py]: @kbrunik took care of this, good catch!

**Pull request merged**